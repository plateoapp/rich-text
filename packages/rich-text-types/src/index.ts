/**
 * Map of all Plate Rich Text Blocks.
 */
export enum BLOCKS {
  Paragraph = 'paragraph',

  HeadingOne = 'heading-one',
  HeadingTwo = 'heading-two',
  HeadingThree = 'heading-three',
  HeadingFour = 'heading-four',
  HeadingFive = 'heading-five',
  HeadingSix = 'heading-six',

  BulletedList = 'bulleted-list',
  NumberedList = 'numbered-list',
  ListItem = 'list-item',

  Quote = 'quote',

  Image = 'image',
}

/**
 * Map of all Plateo Rich Text Inlines.
 */
export enum INLINES {
  Link = 'link',
}

/**
 * Map of all Plateo Rich Text Marks.
 */
export enum MARKS {
  Bold = 'bold',
  Italic = 'italic',
  Underline = 'underline',
  Code = 'code',
}

/**
 * Element are a type af Nodes in a Plateo document. Element can contain other nodes
 * or text nodes. Element nodes can either be `Blocks` or `Inlines`.
 */
export interface Element {
  children: Array<Text | Block | Inline>;
}

/**
 * Text represent the node that can contain actual text along with formatting properties like bold, underline etc.
 */
export interface Text {
  text: string;
  [MARKS.Bold]?: boolean;
  [MARKS.Italic]?: boolean;
  [MARKS.Underline]?: boolean;
  [MARKS.Code]?: boolean;
}

export interface Block extends Element {
  type: BLOCKS;
}

export interface Inline extends Element {
  type: INLINES;
}

export type Document = Array<Text | Block | Inline>;

export interface Paragraph extends Block {
  type: BLOCKS.Paragraph;
}

export interface HeadingOne extends Block {
  type: BLOCKS.HeadingOne;
}

export interface HeadingTwo extends Block {
  type: BLOCKS.HeadingTwo;
}

export interface HeadingThree extends Block {
  type: BLOCKS.HeadingThree;
}

export interface HeadingFour extends Block {
  type: BLOCKS.HeadingFour;
}

export interface HeadingFive extends Block {
  type: BLOCKS.HeadingFive;
}

export interface HeadingSix extends Block {
  type: BLOCKS.HeadingSix;
}

export interface BulletedList extends Block {
  type: BLOCKS.BulletedList;
}

export interface NumberedList extends Block {
  type: BLOCKS.NumberedList;
}

export interface ListItem extends Block {
  type: BLOCKS.ListItem;
}

export interface Image extends Block {
  type: BLOCKS.Image;
  target: {
    id: string;
    linkType: 'Asset';
    type: 'Link';
  };
  children: Text[];
}

export interface Link extends Inline {
  type: INLINES.Link;
  url: string;
  children: Text[];
}
