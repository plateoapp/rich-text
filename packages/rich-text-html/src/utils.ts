import { Text } from '@plateo/rich-text-types';

export function isObject(val: any) {
  return val && typeof val === 'object';
}

export function isText(value: any): value is Text {
  return isObject(value) && typeof value.text === 'string';
}
