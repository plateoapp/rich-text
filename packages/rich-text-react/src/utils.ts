import * as React from 'react';
import { Text } from '@plateo/rich-text-types';

export function isObject(val: any) {
  return val && typeof val === 'object';
}

export function isText(value: any): value is Text {
  return isObject(value) && typeof value.text === 'string';
}

export function addKeyToElement(element: React.ReactNode, key: number) {
  if (React.isValidElement(element) && element.key === null) {
    return React.cloneElement(element, { key });
  }

  return element;
}
