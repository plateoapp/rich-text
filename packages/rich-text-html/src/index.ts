import escapeHtml from 'escape-html';

import {
  Block,
  BLOCKS,
  Document,
  Text,
  Inline,
  INLINES,
  MARKS,
  Link,
  Image,
} from '@plateo/rich-text-types';

import { isText } from './utils';

/**
 * Union type of the different types
 */
export type NodeElement = Text | Block | Inline;

export interface Nested {
  (nodes: NodeElement[]): string;
}

export interface Renderer {
  (node: Block | Inline, nested: Nested): string;
}

export interface NodeRenderer {
  [key: string]: Renderer;
}

export interface MarkRenderer {
  [key: string]: (text: string) => string;
}

export interface RenderOptions {
  /**
   * Node renderers
   */
  node?: NodeRenderer;
  /**
   * Mark renderers
   */
  mark?: MarkRenderer;
}

interface Renderers {
  node: NodeRenderer;
  mark: MarkRenderer;
}

const defaultNodeRenderer: NodeRenderer = {
  [BLOCKS.Paragraph]: (node, nested) => {
    return `<p>${nested(node.children)}</p>`;
  },
  [BLOCKS.HeadingOne]: (node, nested) => {
    return `<h1>${nested(node.children)}</h1>`;
  },
  [BLOCKS.HeadingTwo]: (node, nested) => {
    return `<h2>${nested(node.children)}</h2>`;
  },
  [BLOCKS.HeadingThree]: (node, nested) => {
    return `<h3>${nested(node.children)}</h3>`;
  },
  [BLOCKS.HeadingFour]: (node, nested) => {
    return `<h4>${nested(node.children)}</h4>`;
  },
  [BLOCKS.HeadingFive]: (node, nested) => {
    return `<h5>${nested(node.children)}</h5>`;
  },
  [BLOCKS.HeadingSix]: (node, nested) => {
    return `<h6>${nested(node.children)}</h6>`;
  },
  [BLOCKS.BulletedList]: (node, nested) => {
    return `<ul>${nested(node.children)}</ul>`;
  },
  [BLOCKS.NumberedList]: (node, nested) => {
    return `<ol>${nested(node.children)}</ol>`;
  },
  [BLOCKS.ListItem]: (node, nested) => {
    return `<li>${nested(node.children)}</li>`;
  },
  [BLOCKS.Quote]: (node, nested) => {
    return `<blockquote>${nested(node.children)}</blockquote>`;
  },
  [BLOCKS.Image]: node => {
    return imageBlock(node as Image);
  },
  [INLINES.Link]: (node, nested) => {
    return inlineLink(node as Link, nested);
  },
};

const defaultMarkRenderer: MarkRenderer = {
  [MARKS.Bold]: text => `<strong>${text}</strong>`,
  [MARKS.Italic]: text => `<i>${text}</i>`,
  [MARKS.Underline]: text => `<u>${text}</u>`,
  [MARKS.Code]: text => `<code>${text}</code>`,
};

function imageBlock(node: Image) {
  return `<span>${node.type} id: ${node.target.id}</span>`;
}

function inlineLink(node: Link, nested: Nested) {
  return `<a href="${escapeHtml(node.url)}">${nested(node.children)}</a>`;
}

// TODO: Support multiple marks on a single text block.
function renderText(node: Text, mark: MarkRenderer) {
  const textValue = escapeHtml(node.text);
  let value = textValue;

  if (node.bold) {
    value = mark[MARKS.Bold](textValue);
  }

  if (node.italic) {
    value = mark[MARKS.Italic](textValue);
  }

  if (node.underline) {
    value = mark[MARKS.Underline](textValue);
  }

  if (node.code) {
    value = mark[MARKS.Code](textValue);
  }

  return value;
}

function serializeRichText(node: NodeElement, renderers: Renderers) {
  if (isText(node)) {
    return renderText(node, renderers.mark);
  }

  const nested: Nested = nodes => {
    return nodesToHtmlString(nodes, renderers);
  };

  if (!node.type || !renderers.node[node.type]) {
    return '';
  }

  return renderers.node[node.type](node, nested);
}

function nodesToHtmlString(nodes: NodeElement[], renderers: Renderers): string {
  return nodes
    .map((node: NodeElement) => serializeRichText(node, renderers))
    .join('');
}

/**
 * Serialize a Plateo Rich Text Document into an html string.
 */
export function convertDocumentToHtmlString(
  richTextJSON: Document,
  renderOptions: RenderOptions = {}
) {
  if (!richTextJSON || !Array.isArray(richTextJSON)) {
    return '';
  }

  return nodesToHtmlString(richTextJSON, {
    node: {
      ...defaultNodeRenderer,
      ...renderOptions.node,
    },
    mark: {
      ...defaultMarkRenderer,
      ...renderOptions.mark,
    },
  });
}
