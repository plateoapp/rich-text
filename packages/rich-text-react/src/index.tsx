import * as React from 'react';

import {
  Block,
  BLOCKS,
  Document,
  Text,
  Inline,
  INLINES,
  MARKS,
  Link,
  Image,
} from '@plateo/rich-text-types';

import { isText, addKeyToElement } from './utils';

/**
 * Union type of the different types
 */
export type NodeElement = Text | Block | Inline;

export interface Nested {
  (nodes: NodeElement[]): React.ReactNode;
}

export interface Renderer {
  (node: Block | Inline, nested: Nested): React.ReactNode;
}

export interface NodeRenderer {
  [key: string]: Renderer;
}

export interface MarkRenderer {
  [key: string]: (text: string) => React.ReactNode;
}

export interface RenderOptions {
  /**
   * Node renderers
   */
  node?: NodeRenderer;
  /**
   * Mark renderers
   */
  mark?: MarkRenderer;
}

interface Renderers {
  node: NodeRenderer;
  mark: MarkRenderer;
}

const defaultNodeRenderer: NodeRenderer = {
  [BLOCKS.Paragraph]: (node, nested) => {
    return <p>{nested(node.children)}</p>;
  },
  [BLOCKS.HeadingOne]: (node, nested) => {
    return <h1>{nested(node.children)}</h1>;
  },
  [BLOCKS.HeadingTwo]: (node, nested) => {
    return <h2>{nested(node.children)}</h2>;
  },
  [BLOCKS.HeadingThree]: (node, nested) => {
    return <h3>{nested(node.children)}</h3>;
  },
  [BLOCKS.HeadingFour]: (node, nested) => {
    return <h4>{nested(node.children)}</h4>;
  },
  [BLOCKS.HeadingFive]: (node, nested) => {
    return <h5>{nested(node.children)}</h5>;
  },
  [BLOCKS.HeadingSix]: (node, nested) => {
    return <h6>{nested(node.children)}</h6>;
  },
  [BLOCKS.BulletedList]: (node, nested) => {
    return <ul>{nested(node.children)}</ul>;
  },
  [BLOCKS.NumberedList]: (node, nested) => {
    return <ol>{nested(node.children)}</ol>;
  },
  [BLOCKS.ListItem]: (node, nested) => {
    return <li>{nested(node.children)}</li>;
  },
  [BLOCKS.Quote]: (node, nested) => {
    return <blockquote>{nested(node.children)}</blockquote>;
  },
  [BLOCKS.Image]: node => {
    return imageBlock(node as Image);
  },
  [INLINES.Link]: (node, nested) => {
    return inlineLink(node as Link, nested);
  },
};

const defaultMarkRenderer: MarkRenderer = {
  [MARKS.Bold]: text => <strong>{text}</strong>,
  [MARKS.Italic]: text => <i>{text}</i>,
  [MARKS.Underline]: text => <u>{text}</u>,
  [MARKS.Code]: text => <code>{text}</code>,
};

function imageBlock(node: Image) {
  return (
    <span key={node.target.id}>
      {node.type} id: {node.target.id}
    </span>
  );
}

function inlineLink(node: Link, nested: Nested) {
  return <a href={node.url}>{nested(node.children)}</a>;
}

// TODO: Support multiple marks on a single text block.
function renderText(node: Text, mark: MarkRenderer): string | React.ReactNode {
  const textValue = node.text;
  let value: string | React.ReactNode = textValue;

  if (node.bold) {
    value = mark[MARKS.Bold](textValue);
  }

  if (node.italic) {
    value = mark[MARKS.Italic](textValue);
  }

  if (node.underline) {
    value = mark[MARKS.Underline](textValue);
  }

  if (node.code) {
    value = mark[MARKS.Code](textValue);
  }

  return value;
}

function serializeRichText(
  node: NodeElement,
  renderers: Renderers
): React.ReactNode {
  if (isText(node)) {
    return renderText(node, renderers.mark);
  }

  const nested: Nested = nodes => {
    return nodesToReactComponents(nodes, renderers);
  };

  if (!node.type || !renderers.node[node.type]) {
    return null;
  }

  return renderers.node[node.type](node, nested);
}

function nodesToReactComponents(
  nodes: NodeElement[],
  renderers: Renderers
): React.ReactNode {
  return nodes.map((node: NodeElement, index) =>
    addKeyToElement(serializeRichText(node, renderers), index)
  );
}

/**
 * Serialize a Plateo Rich Text Document into React components.
 */
export function convertDocumentToReactComponents(
  richTextJSON: Document,
  renderOptions: RenderOptions = {}
) {
  if (!richTextJSON || !Array.isArray(richTextJSON)) {
    return null;
  }

  return nodesToReactComponents(richTextJSON, {
    node: {
      ...defaultNodeRenderer,
      ...renderOptions.node,
    },
    mark: {
      ...defaultMarkRenderer,
      ...renderOptions.mark,
    },
  });
}
