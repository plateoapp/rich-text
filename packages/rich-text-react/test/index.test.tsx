import * as React from 'react';

import {
  Document,
  BLOCKS,
  Block,
  Inline,
  MARKS,
  INLINES,
} from '@plateo/rich-text-types';

import { convertDocumentToReactComponents, Nested } from '../src/index';

describe('convertDocumentToReactComponents', () => {
  it('returns empty array when given an empty document', () => {
    const document: unknown = [];

    expect(convertDocumentToReactComponents(document as Document)).toEqual([]);
  });

  it('can handle undefined documents', () => {
    const document: unknown = undefined;
    expect(convertDocumentToReactComponents(document as Document)).toEqual(
      null
    );
  });

  it('renders document with default node renderer', () => {
    const documents: Document[] = [
      [
        {
          type: BLOCKS.HeadingOne,
          children: [
            {
              text: 'I work',
            },
          ],
        },
      ],
      [
        {
          type: BLOCKS.Paragraph,
          children: [
            {
              text: 'Nice Paragraph',
            },
          ],
        },
      ],
    ];

    documents.forEach(node => {
      expect(convertDocumentToReactComponents(node)).toMatchSnapshot();
    });
  });

  it('renders document with default mark renderer', () => {
    const document: Document = [
      {
        type: BLOCKS.Paragraph,
        children: [
          {
            text: 'Nice Paragraph',
            bold: true,
          },
        ],
      },
      {
        type: BLOCKS.Paragraph,
        children: [
          {
            text: 'Nice Paragraph',
            italic: true,
          },
        ],
      },
      {
        type: BLOCKS.Paragraph,
        children: [
          {
            text: 'Nice Paragraph',
            underline: true,
          },
        ],
      },
      {
        type: BLOCKS.Paragraph,
        children: [
          {
            text: 'Nice Paragraph',
            code: true,
          },
        ],
      },
    ];

    expect(convertDocumentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders document with a custom node renderer', () => {
    const renderOptions = {
      node: {
        [BLOCKS.Paragraph]: (node: Block | Inline, nested: Nested) => (
          <div>{nested(node.children)}</div>
        ),
      },
    };

    const documents: Document[] = [
      [
        {
          type: BLOCKS.Paragraph,
          children: [
            {
              text: 'Nice Paragraph',
              bold: true,
            },
          ],
        },
      ],
      [
        {
          type: BLOCKS.Paragraph,
          children: [
            {
              text: 'Nice Paragraph',
              underline: true,
            },
          ],
        },
      ],
    ];

    documents.forEach(node => {
      expect(
        convertDocumentToReactComponents(node, renderOptions)
      ).toMatchSnapshot();
    });
  });

  it('renders document with a custom mark renderer', () => {
    const renderOptions = {
      mark: {
        [MARKS.Underline]: (text: string) => <strong>{text}</strong>,
      },
    };

    const document: Document = [
      {
        type: BLOCKS.Paragraph,
        children: [
          {
            text: 'Nice Paragraph',
            underline: true,
          },
        ],
      },
    ];

    expect(
      convertDocumentToReactComponents(document, renderOptions)
    ).toMatchSnapshot();
  });

  it('renders asset link', () => {
    const document = [
      {
        type: BLOCKS.Image,
        children: [
          {
            text: '',
          },
        ],
        target: {
          id: '070fec38-1038-4b2d-ae67-06208c2dee50',
          linkType: 'Asset',
          type: 'Link',
        },
      },
    ];

    expect(convertDocumentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders link', () => {
    const document = [
      {
        type: BLOCKS.Paragraph,
        children: [
          {
            text: 'on this page: ',
          },
          {
            type: INLINES.Link,
            children: [
              {
                text: 'Google',
              },
            ],
            url: 'https://google.com',
          },
          {
            text: ' some text',
          },
        ],
      },
    ];

    expect(convertDocumentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders bulleted list', () => {
    const document = [
      {
        type: BLOCKS.BulletedList,
        children: [
          {
            type: BLOCKS.ListItem,
            children: [
              {
                text: 'some text',
              },
            ],
          },
        ],
      },
    ];

    expect(convertDocumentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders numbered list', () => {
    const document = [
      {
        type: BLOCKS.NumberedList,
        children: [
          {
            type: BLOCKS.ListItem,
            children: [
              {
                text: 'some text',
              },
            ],
          },
        ],
      },
    ];

    expect(convertDocumentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders block quote', () => {
    const document = [
      {
        type: BLOCKS.Quote,
        children: [
          {
            text: 'some quote',
          },
        ],
      },
    ];

    expect(convertDocumentToReactComponents(document)).toMatchSnapshot();
  });
});
