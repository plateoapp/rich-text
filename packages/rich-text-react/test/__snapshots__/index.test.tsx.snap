// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertDocumentToReactComponents renders asset link 1`] = `
Array [
  <span>
    image
     id: 
    070fec38-1038-4b2d-ae67-06208c2dee50
  </span>,
]
`;

exports[`convertDocumentToReactComponents renders block quote 1`] = `
Array [
  <blockquote>
    some quote
  </blockquote>,
]
`;

exports[`convertDocumentToReactComponents renders bulleted list 1`] = `
Array [
  <ul>
    <li>
      some text
    </li>
  </ul>,
]
`;

exports[`convertDocumentToReactComponents renders document with a custom mark renderer 1`] = `
Array [
  <p>
    <strong>
      Nice Paragraph
    </strong>
  </p>,
]
`;

exports[`convertDocumentToReactComponents renders document with a custom node renderer 1`] = `
Array [
  <div>
    <strong>
      Nice Paragraph
    </strong>
  </div>,
]
`;

exports[`convertDocumentToReactComponents renders document with a custom node renderer 2`] = `
Array [
  <div>
    <u>
      Nice Paragraph
    </u>
  </div>,
]
`;

exports[`convertDocumentToReactComponents renders document with default mark renderer 1`] = `
Array [
  <p>
    <strong>
      Nice Paragraph
    </strong>
  </p>,
  <p>
    <i>
      Nice Paragraph
    </i>
  </p>,
  <p>
    <u>
      Nice Paragraph
    </u>
  </p>,
  <p>
    <code>
      Nice Paragraph
    </code>
  </p>,
]
`;

exports[`convertDocumentToReactComponents renders document with default node renderer 1`] = `
Array [
  <h1>
    I work
  </h1>,
]
`;

exports[`convertDocumentToReactComponents renders document with default node renderer 2`] = `
Array [
  <p>
    Nice Paragraph
  </p>,
]
`;

exports[`convertDocumentToReactComponents renders link 1`] = `
Array [
  <p>
    on this page: 
    <a
      href="https://google.com"
    >
      Google
    </a>
     some text
  </p>,
]
`;

exports[`convertDocumentToReactComponents renders numbered list 1`] = `
Array [
  <ol>
    <li>
      some text
    </li>
  </ol>,
]
`;
